app/controllers/application_controller.rb

```diff
@@ -20,12 +19,21 @@ class ApplicationController < ActionController::Base
   end
 
   around_action :switch_locale
-
   def switch_locale(&action)
-    locale = params[:locale] || I18n.default_locale
+    locale = if params[:locale] == 'ja'
+      params[:locale]
+    elsif extract_locale_from_accept_language_header == 'ja'
+      redirect_to(request.original_url + "?locale=ja") and return
+    else
+      I18n.default_locale
+    end
     I18n.with_locale(locale, &action)
   end
 
+  def default_url_options
```

Comment:
Does this mean that locale will be passed through to every single link? I'm not sure we want that and it probably would be better to read from the header